### ISSUE TITLE
UltraFastBERT FFF Layer performs worse than MLP

### ISSUE THREAD

**Issue by catid** [pbelcak/UltraFastBERT#12]
I've been trying to reproduce your positive results for the FFF layer structure. To simplify the comparison I've been using CIFAR-10 as a proxy problem.

Over the past week I put together a training framework for CIFAR-10 with a baseline transformer model (vit_tiny with mlp_dim=256). I've then introduced a number of variants of the transformer model using your UltraFastBERT implementation of FFF, some tweaks to it, and a community version of FFF written from scratch. The results are here:

![image](https://github.com/pbelcak/UltraFastBERT/assets/252193/c895aa99-daf3-4c53-9470-922918b7b370)

Code for this experiment is here: https://github.com/catid/cifar10deepspeed

So far we have yet to see the FFF layer improve upon a small mlp_dim=16 FFN network. The conditional computation does not seem to improve the network's ability to generalize to the validation set.

I'm currently suspecting that the UltraFastBERT result can be improved by replacing the FFF layers with a MLP layer that is with mlp_dim=16, which is obviously much smaller and easier to train/evaluate than a FFF layer.

**Comment by catid** [comment]
As mentioned in https://github.com/pbelcak/UltraFastBERT/issues/9 the scripts in this repo do not reproduce the paper results.  So, I do not have a way to evaluate this modification in context of the FFF results at this time.  If you have working code somewhere please perform this experiment:
(1) Train UltraFastBERT again.
(2) Replace the FFF layer with MLP with hidden dimension 16 or 32 or something else small.

Please let us know which one performs better.
